_posts/2023-02-13-elfsign.md

```diff
@@ -0,0 +1,159 @@
+(which is a topic of its own). Even if you create a "fat", or "shaded" JAR, it is possible to understand what
+really ended up in your "binary".
+
+And, others can do the same. Rust for example allows one use `cargo auditable`, to tap into the compilation process,
+and record the actual dependencies which go into a binary. On Linux, the binary will be an ELF file, which then
+contains the dependency information from the compilation process. And Go can to the same.
+
+## Null and void
+
+But, if someone can write dependency information into a binary, then someone else can also overwrite it. So
+unless you protect the binary against modifications, this information isn't really trustworthy. Taking a look at
+the Helm release page, you will find SHA based checksums. Isn't that enough?
+
+No, not really. Because if someone can alter the binary on the download page, the attacker can also swap out the
+SHA checksum file. The checksum, or digest, really isn't more than a checksum. By its own, it doesn't protect much.
+
+If however, you encrypt the digest with a private key that only the author knows, and you publish the public key
+part, then this becomes a "signature". And this is good enough to prove, that only the person how knew the private
+key, could have signed the binary. And if we can trust this person to create the correct SBOM and binary, and keep
+the private key secure, we are good.
+
+## Usability
+
+Or, not! In the Maven world JARs get signed for quite a while. Everyone uploading JARs to Maven Central needs
+to sign their JAR with GPG. And I guess most people never validated a JAR during a build.
+
+On the other side, the Eclipse IDE (as well as some other Java applications), did JAR file validation for quite a
+while. Whenever you install a plugin, it cryptographically validates the JAR. And it's easy, as the JAR file itself
+is signed, and the signature is part of the JAR file. As part of the build process in the Eclipse Foundation
+build system, JARs which got created by the build, can automatically get signed. No additional files needed,
+and only the actual build output is considered, no guessing of dependencies.
+
+From a user perspective, the IDE automatically checks signatures, and only bothers the user if an issue was found.
+The user can override, because the idea is to give the final authority to the user.
+
+## Back to binaries
+
+Now, just assume we could do the same with (ELF) binaries. In fact this was possible a while back, Solaris had some
+tools to sign ELF binaries. That doesn't help on modern Linux systems. But with a bit of Rust code, it was possible
+to create [elfsign](https://github.com/ctron/elfsign). The idea is simple:
+
+* Create a digest of all relevant ELF sections and headers
+* Sign this digest
+* Add the signature to the ELF binary as a new section
+
+## Solvable downsides
+
+The first downside might be that people are afraid of signed binaries. Windows and macOS do this for a while,
```

Comment:
Suggestion: "Windows and macOS have been doing this for a while, and it happens that signatures get in the way of the user running a binary"